src/components/search/InputWithSuggester.tsx

```diff
@@ -135,50 +138,90 @@ const InputWithSuggester: FC<InputWithSuggesterProps> = ({
-          });
-          phrasesSuggestions.forEach((phrase: string) => {
-            options.push({ text: phrase, group: OptionGroup.PHRASE });
+  const refreshOptions = useCallback(
+    async (inputValue: string) => {
+      // setPendingSearch to true to prevent doing search before refreshing options is finished
+      setPendingSearch(true);
+      try {
+        const currentState: ParameterState = getComponentState(
+          store.getState()
+        );
+        dispatch(fetchSuggesterOptions(createSuggesterParamFrom(currentState)))
+          .unwrap()
+          .then((data) => {
+            const categorySuggestions = new Set<string>(
+              data.category_suggestions
+            );
+            const phrasesSuggestions = new Set<string>(
+              data.record_suggestions.suggest_phrases
+            );
+
+            // Get the intersection of categorySuggestions and phrasesSuggestions
+            const commonSuggestions = [...categorySuggestions].filter(
+              (item) => {
+                return phrasesSuggestions.has(item);
+              }
+            );
+
+            // check if the current inputValue is a common key or not, then dispatch updateCommonKey
+            if (
+              commonSuggestions.some(
+                (item) => item.toLowerCase() === inputValue.toLowerCase()
+              )
+            ) {
+              dispatch(updateCommonKey(inputValue));
+            } else {
+              dispatch(updateCommonKey(""));
+            }
+
+            // Create array of all unique suggestions
+            const allSuggestions = [
+              ...new Set([
+                ...commonSuggestions,
+                ...categorySuggestions,
+                ...phrasesSuggestions,
+              ]),
+            ];
+
+            // Sort suggestions by relevance
+            const sortedSuggestions = sortBy(allSuggestions, [
+              // Exact match first
+              (s) => s.toLowerCase() !== inputValue.toLowerCase(),
+              // Then by whether it starts with the input
+              (s) => !s.toLowerCase().startsWith(inputValue.toLowerCase()),
+              // Then by the index where the input appears in the suggestion
+              (s) => s.toLowerCase().indexOf(inputValue.toLowerCase()),
+              // Finally by length (shorter suggestions first)
+              (s) => s.length,
+            ]);
+
+            // Create sorted options array
+            const options: OptionType[] = sortedSuggestions.map(
+              (suggestion) => {
+                if (commonSuggestions.includes(suggestion)) {
+                  return { text: suggestion, group: OptionGroup.COMMON };
+                } else if (categorySuggestions.has(suggestion)) {
+                  return { text: suggestion, group: OptionGroup.CATEGORY };
+                } else {
+                  return { text: suggestion, group: OptionGroup.PHRASE };
+                }
+              }
+            );
```

Comment:
all good but for readability, you might want to extract generic functions from this block, they extracted methods stay outside the `.then()`, and **maybe** reusable by different components in future. e.g.

```javascript
const sortStringList = (stringList: string[], inputValue: string) => {
...
}
```

then call it (line 187)

```javascript
const sortedSuggestions = sortStringList([...allSuggestions], inputValue);
```

Reply:
Thanks Viet! I have extracted a util function and move to utils folder